Add DevicePath functions/utilities

csssuf opened the pull request:
Primary changes:
* `DevicePathUtilities` for manipulating `DevicePathProtocol` structs
* `DevicePathFromText` for parsing strings into `DevicePathProtocol` structs
* `create_file_device_node` for creating `DevicePathProtocol` structs for files

csssuf committed:
device_path: use &str for DevicePathToTextProtocol

Instead of using *const u16s and Console.write_raw() in
DevicePathToTextProtocol, use &str and write() instead.

csssuf committed:
util/device_path: add create_file_device_node

Add a function to allocate and create a DevicePathProtocol struct
containing a filename.

csssuf committed:
protocol/device_path: add DevicePathFromText

csssuf committed:
protocol/device_path: add DevicePathUtilities

csssuf committed:
lib: publicly export CVoid

CVoid is needed for passing to some functions.

bgilbert commented on src/util/device_path.rs:
```diff
@@ -0,0 +1,44 @@
+
+                        for i in 0..filename_len as isize{
+                            unsafe {
+                                *file_name_ptr.offset(i) = *filename.as_ptr().offset(i) as u16;
```
Should be setting `file_name_ptr` from `filename_ptr`, not `filename.as_ptr()`.

Also maybe some better names?

bgilbert commented on src/protocol/device_path.rs:
```diff
@@ -128,6 +131,30 @@ impl DevicePathToTextProtocol {
+impl DevicePathFromTextProtocol {
+    pub fn text_to_device_path_node(&self, path: &str) -> Result<&DevicePathProtocol, Status> {
+        str_to_utf16_ptr(path)
+            .map(|utf16_str| unsafe { &*((self.text_to_device_path_node)(utf16_str)) })
```
Is `utf16_str` leaked?

bgilbert commented on src/protocol/device_path.rs:
```diff
@@ -128,6 +131,30 @@ impl DevicePathToTextProtocol {
+
+    pub fn text_to_device_path(&self, path: &str) -> Result<&DevicePathProtocol, Status> {
+        str_to_utf16_ptr(path)
+            .map(|utf16_str| unsafe { &*((self.text_to_device_path)(utf16_str)) })
```
Likewise.

bgilbert commented on src/util/device_path.rs:
```diff
@@ -0,0 +1,44 @@
+            .boot_services()
+            .locate_protocol::<DevicePathUtilitiesProtocol>(0 as *const CVoid)
+            .and_then(|utilities| {
+                utilities.create_device_node(4, 4, node_size_bytes as u16)
```
Is `4` worth a couple of constants?

csssuf replied:
Yes, probably. Realistically the best thing to do is to flesh out all the DevicePath types and subtypes from the spec, I was just a little hesitant to do that since we're not using many of them.

csssuf committed:
protocol/device_path: add all device path type enums

Add enums for all device path types as laid out in the UEFI spec. Also,
add implementations of Into<u8> for each.

csssuf commented:
Updated with requested changes.

bgilbert commented on src/util/device_path.rs:
```diff
@@ -0,0 +1,44 @@
+
+                        for i in 0..filename_len as isize{
+                            unsafe {
+                                *node_filename_ptr.offset(i) = *filename_ptr.offset(i) as u16;
```
Is `as u16` redundant here?

bgilbert commented on src/protocol/device_path.rs:
```diff
@@ -29,24 +31,141 @@ pub enum DevicePathTypes {
+
+#[allow(non_camel_case_types)]
+#[repr(u8)]
+pub enum MessagingSubTypes {
```
Please either sort, or leave a comment explaining the sort order.

csssuf replied:
Sorted by value, since the spec order doesn't seem to have any rhyme or reason.

bgilbert commented on src/protocol/device_path.rs:
```diff
@@ -29,24 +31,141 @@ pub enum DevicePathTypes {
+    IPv6 = 0xD,
+    UART = 0xE,
+    USBClass = 0xF,
+    // The EFI specification calls this type "1394", but that won't work as an enum variant for
```
Nope.

csssuf replied:
Oops, yep. Thanks.
